introduce migrate:fresh command

thyseus opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | N/A

thyseus committed:
introduce migrate:fresh command

pana1990 commented:
@thyseus What difference does it execute the following command?

```bash
./yii migrate/redo all
```

thyseus commented:
./yii migrate/redo all tries to replay all migrations, but does not truncate the existing database beforehand.

samdark commented on framework/console/controllers/MigrateController.php:
```diff
@@ -273,6 +273,34 @@ protected function addMigrationHistory($version)
+        // First drop all foreign keys,
+        foreach ($schemas as $schema) {
+            if ($schema->foreignKeys) {
+                foreach ($schema->foreignKeys as $name => $foreign_key) {
```
Please use camel case: `$foreignKey`.

samdark commented on framework/CHANGELOG.md:
```diff
@@ -15,6 +15,7 @@ Yii Framework 2 Change Log
 - Bug #14533: Fixed `yii\validators\ExistValidator` and `yii\validators\UniqueValidator` throw exception in case they are set for `yii\db\ActiveRecord` with `$targetClass` pointing to NOSQL ActiveRecord (klimov-paul)
 - Bug #14449: Fix PHP 7.2 compatibility bugs and add explicit closure support in `yii\base\Application` (dynasource)
 - Bug #7890: Allow `migrate/mark` to mark history at the point of the base migration (cebe)
+- Enh #14664: Add migrate/fresh command to reset database and restart migration (idea from laravel 5.5) (thyseus)
```
No need to mention Laravel here :)

samdark reviewed (approved):
Overall I like it. Please fix two minor things and I'll merge it.

thyseus commented:
@samdark style improved. Should i add some lines to the documentation?

thyseus commented:
Oh, and should i add an check that this is only possible on YII_ENV != 'prod' ?

cebe commented on framework/console/controllers/MigrateController.php:
```diff
@@ -273,6 +273,35 @@ protected function addMigrationHistory($version)
 
     /**
      * @inheritdoc
```
there is no parent method to inherit docs from. `refreshDatabase` method should be implemented in BaseMigrateController and print something like `'This command is not implemented in ' . get_class($this)`

cebe commented:
> @samdark style improved. Should i add some lines to the documentation?

yes

> Oh, and should i add an check that this is only possible on YII_ENV != 'prod' ?

and yes :)

pana1990 commented:
@thyseus thanks for your explain :), you have my like

thyseus commented:
@cebe, @samdark next iteration ready to be reviewed.
@pana1990 You´re welcome !

cebe commented on framework/console/controllers/BaseMigrateController.php:
```diff
@@ -428,6 +428,29 @@ public function actionMark($version)
+     */
+    public function actionFresh()
+    {
+        if (defined('YII_ENV') && YII_ENV == 'prod') {
```
can be simply `if (YII_ENV_PROD)`

cebe commented on framework/console/controllers/BaseMigrateController.php:
```diff
@@ -877,6 +900,13 @@ protected function generateMigrationSourceCode($params)
+     * Truncates the database and reapplies all migrations from the beginning.
+     * @return void
+     */
+    abstract protected function refreshDatabase();
```
abstract function will introduce BC break, extensions that do not implement it, will break. It should be implemented and just print a message that it is not implemented.

cebe replied:
See https://github.com/yiisoft/yii2/pull/14664/files#r133911128

thyseus replied:
ok. now i understand. fixed it.

cebe commented on framework/console/controllers/BaseMigrateController.php:
```diff
@@ -877,6 +900,13 @@ protected function generateMigrationSourceCode($params)
+    /**
+     * Truncates the database and reapplies all migrations from the beginning.
+     * @return void
+     */
```
`@since 2.0.13`

thyseus commented:
Very cool so far! I will add the unit tests on monday when i am back in my office.

thyseus committed:
add test for migrate:fresh command

thyseus commented:
@samdark, @cebe i could not wait until monday :) 

i've added and test and also did an rebase and squashed your commits into mine, so the history looks cool once merged into master.

samdark commented:
Merged. Thank you!

thiagotalma commented on docs/guide/db-migrations.md:
```diff
@@ -788,6 +788,13 @@ yii migrate/redo 3      # redo the last 3 applied migrations
 
+## Refreshing Migrations <span id="refreshing-migrations"></span>
+
+Since Yii 2.0.13 you can reset the whole database and apply all migrations from the beginning.
```
That statement may need a change.
This is not entirely true, only the tables will be erased and the database has other elements like functions and stored procedures.
You only guarantee a "complete reset" using "DROP/CREATE DATABASE".
cc: @samdark

samdark replied:
https://github.com/yiisoft/yii2/commit/a3cd99d77d9748d4edafd28ddecd7d32d7e1a3ee

schmunk42 commented:
What's about renaming this to `reset`? 

Re`fresh`ing does not really indicate the complete data loss IMHO. I'd rather expect something which "fixes" my current db in case something went wrong. `reset` would be more explicit.

thiagotalma commented:
> What's about renaming this to reset?

@schmunk42  https://github.com/yiisoft/yii2/pull/14664#discussion_r148796421
